sky/jobs/scheduler.py

```diff
@@ -0,0 +1,218 @@
+"""Scheduler for managed jobs.
+
+Once managed jobs are added as PENDING to the `spot` table, the scheduler is
+responsible for the business logic of deciding when they are allowed to start,
+and choosing the right one to start.
+
+The scheduler is not its own process - instead, schedule_step() can be called
+from any code running on the managed jobs controller to trigger scheduling of
+new jobs if possible. This function should be called immediately after any state
+change that could result in new jobs being able to start.
+
+The scheduling logic limits the number of running jobs according to two limits:
+1. The number of jobs that can be launching (that is, STARTING or RECOVERING) at
+   once, based on the number of CPUs. (See _get_launch_parallelism.) This the
+   most compute-intensive part of the job lifecycle, which is why we have an
+   additional limit.
+2. The number of jobs that can be running at any given time, based on the amount
+   of memory. (See _get_job_parallelism.) Since the job controller is doing very
+   little once a job starts (just checking its status periodically), the most
+   significant resource it consumes is memory.
+
+There are two ways to interact with the scheduler:
+- Any code that could result in new jobs being able to start (that is, it
+  reduces the number of jobs counting towards one of the above limits) should
+  call schedule_step(), which will best-effort attempt to schedule new jobs.
+- If a running job need to relaunch (recover), it should use schedule_recovery()
+  to obtain a "slot" in the number of allowed starting jobs.
+
+Since the scheduling state is determined by the state of jobs in the `spot`
+table, we must sychronize all scheduling logic with a global lock. A per-job
+lock would be insufficient, since schedule_step() could race with a job
+controller trying to start recovery, "double-spending" the open slot.
+"""
+
+import contextlib
+import os
+import shlex
+import subprocess
+import time
+from typing import Optional, Tuple
+
+import filelock
+import psutil
+
+from sky.jobs import state, constants as managed_job_constants
+from sky.skylet import constants
+from sky.utils import subprocess_utils
+
+# The _MANAGED_JOB_SUBMISSION_LOCK should be held whenever a job transitions to
+# STARTING or RECOVERING, so that we can ensure correct parallelism control.
+_MANAGED_JOB_SUBMISSION_LOCK = '~/.sky/locks/managed_job_submission.lock'
+_ACTIVE_JOB_LAUNCH_WAIT_INTERVAL = 0.5
+
+
+def schedule_step() -> None:
```

Comment:
How about calling it `schedule_new_job()`?

Reply:
Changed to maybe_start_waiting_jobs(). Open to other names - it might be a bit of a misnomer if it schedules an existing job that wants to launch something.